core: Prevent corrupting units from stale alias state on daemon-reload

cdown opened the pull request:
During daemon-reload (or daemon-reexec), when a unit becomes an alias to another unit, deserialising the alias's stale serialised state can corrupt the canonical unit's live runtime state.

Consider this scenario:

1. Before reload:
   - a.service is running
   - b.service was stopped earlier and is dead
   - Both exist as independent units

2. User creates an alias to migrate from b -> a:

   - `ln -s /run/systemd/system/a.service /etc/systemd/system/b.service`

3. daemon-reload triggers serialisation. State file contains both units:
   - a.service -> state=running, cgroup=/system.slice/a.service, PID=1234, ...
   - b.service -> state=dead, cgroup=(empty), no PIDs, ...

4. During deserialisation:
   - Processes a.service: loads Unit A, deserialises -> state=RUNNING
   - Processes b.service: manager_load_unit() detects symlink, returns Unit A
   - unit_deserialize_state(Unit A, ...) overwrites with b's dead state

5. The result is that:
   - Unit A incorrectly shows state=dead despite PID 1234 still running
   - If a.service has Upholds= dependents, catch-up logic sees a.service should be running but is dead
   - systemd starts a.service again -> PID 5678
   - Two instances run: PID 1234 (left-over) and PID 5678 (new)

This bug is deterministic when serialisation orders a.service before b.service.

The root cause is that manager_deserialize_one_unit() calls manager_load_unit(name, &u) which resolves aliases via unit_follow_merge(), returning the canonical Unit object. However, the code doesn't distinguish between two cases when u->id differs from the requested name from the state file. In the corruption case, we're deserialising an alias entry and unit_deserialize_state() blindly overwrites the canonical unit's fields with stale data from the old, independent unit. The serialised b.service then overwrites Unit A's correct live state.

This commit first scans the serialised unit names, then adds a check after manager_load_unit():

    if (!streq(u->id, name) && set_contains(serialized_units, u->id))
        ...

This detects when the loaded unit's canonical ID (u->id) differs from the serialised name, indicating the name is now an alias for a different unit and the canonical unit also has its own serialised state entry.

If the canonical unit does not have its own serialised state entry, we keep the state entry. That handles cases where the old name is really just a rename, and thus the old name is the only serialised state for the unit. In that case there is no bug, because there is no separate canonical state entry for the stale alias entry to overwrite.

Skipping is safe because:

1. The canonical unit's own state entry will be correctly deserialised regardless of order. This fix only prevents other stale alias entries from corrupting it.
2. unit_merge() has already transferred the necessary data. When b.service became an alias during unit loading, unit_merge() already migrated dependencies and references to the canonical unit.
3. After merging, the alias doesn't have its own runtime state. The serialised data represents b.service when it was independent, which is now obsolete once the canonical unit also has its own serialised entry.
4. All fields are stale. unit_deserialize_state() would overwrite state, timestamps, cgroup paths, pids, etc. There's no scenario where we want this data applied on top of the canonical unit's own serialised state.

This fix also correctly handles unit precedence. For example, imagine this scenario:

1. `b.service` is a valid, running unit defined in `/run`.
2. The sysadmin creates `ln -s .../a.service /etc/.../b.service`.
3. On reload, the new symlink in `/etc` overrides the unit in `/run`.

The new perspective from the manager side is that `b.service` is now an alias for `a.service`.

In this case, systemd correctly abandons the old b.service unit, because that's the intended general semantics of unit file precedence. We also do that in other cases, like when a unit file in /etc/systemd/system/ masks a vendor-supplied unit file in /lib/systemd/system/, or when an admin uses systemctl mask to explicitly disable a unit.

In all these scenarios, the configuration with the highest precedence (in /etc/) is treated as the new source of truth. The old unit's definition is discarded, and its running processes are (correctly) abandoned. In that respect we are not doing anything new here.

Some may ask why we shouldn't just ignore the symlink if we think this case will come up. I think there are multiple very strong reasons not to do so:

1. It violates unit precedence. The unit design is built on a strict precedence list. When an admin puts any file in /etc, they are intentionally overriding everything else. If manager_load_unit were to "ignore" this file based on runtime state, it would break this fundamental precedent.
2. It makes daemon-reload stateful. daemon-reload is supposed to be a simple, stateless operation, basically to read the files on disk and apply the new configuration. But doing this would make daemon-reload stateful, because we'd have to read the files on disk, but cross-reference the current runtime state, and...  maybe ignore some files. This is complex and unpredictable.
3. It also completely ignores the user intent. The admin clearly has tried to replace the old service with an alias. Ignoring their instruction is the opposite of what they want.

Fixes: https://github.com/systemd/systemd/issues/38817
Fixes: https://github.com/systemd/systemd/issues/37482

cdown commented on src/core/unit-serialize.c:
```diff
@@ -405,6 +406,98 @@ int unit_deserialize_state_skip(FILE *f) {
+                if (r == 0 || !key)
+                        continue;
+
+                if (STR_IN_SET(key,
```
Having a list of these here is not ideal obviously. I'm happy to do some refactor but given these are spread all over the place I think it's probably not realistic. Let me know what you think.

yuwata replied:
Maybe better to add debugging logs about remaining fds in the fdset after manager_deserialize(), and check it in the integration test? We already log about serialized fds at the beginning of manager_deserialize(). Something similar?

cdown replied:
It looks like some FDs are expected to remain after manager_deserialize(), I think bus FDs, for example, are one of those.

So even with logs, how should this be tested? I'm concerned about fragility. The number of remaining FDs seems not expected to be zero in all cases. Over time we may also have different expected counts. What's your suggested approach?

cdown replied:
@yuwata Any updated thoughts on this? :-)

yuwata replied:
Sorry for late response.

> It looks like some FDs are expected to remain after manager_deserialize(), I think bus FDs, for example, are one of those.
>
> So even with logs, how should this be tested? I'm concerned about fragility. The number of remaining FDs seems not expected to be zero in all cases. Over time we may also have different expected counts. What's your suggested approach?

I thought all FDs should be consumed by deserialization. Not sure what 'bus FDs' means here. Have you checked how many FDs are still in the FDSet after deserialization? What kind of FDs are still there?

BTW, at least on daemon-reload, remaining FDs are anyway closed. And, even daemon-reexec, remaining FDs are closed after manager_distribute_fds(). Hence, even if the list here is incomplete or becomes outdated in a future, that should not be a big problem, I guess.

keszybz replied:
Right, so we close the fds here:
https://github.com/systemd/systemd/blob/13a74f350e0629bedb9e3dbcb53cbcff365af699/src/core/main.c#L3428
So I think we should add a message there. Maybe even at warning level. It's iffy if there are any fds left-over.

cdown commented on src/core/manager-serialize.c:
```diff
@@ -202,6 +202,37 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
+                 * before the symlink existed), but b.service now resolves to a.service. We must discard
+                 * b.service's stale dead state to preserve a.service's running state.
+                 */
+                log_warning("Unit file for '%s' was overridden by a symlink to '%s'. Skipping stale state of old unit. Any processes from the overridden unit are now abandoned!",
```
Open to other suggestions. Usually we are relatively benign with the messages, but I think we should make it clear that something quite bad has happened here.

yuwata replied:
Please mention that the log message is checked in TEST-07-PID1.alias-corruption.sh, so the test case may need to be adjusted when the log message is updated.

yuwata commented on src/core/manager-serialize.c:
```diff
@@ -202,6 +202,37 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
+                            name,
+                            u->id);
+
+                r = unit_deserialize_state_discard(f, fds);
```
Simply, 
```c
return unit_deserialize_state_discard(...);
```

yuwata commented on src/core/unit-serialize.c:
```diff
@@ -405,6 +406,98 @@ int unit_deserialize_state_skip(FILE *f) {
+
+                        fd = deserialize_fd(fds, val);
+                        if (fd >= 0)
+                                safe_close(fd);
```
safe_close() gracefully handle negative errno:
```c
safe_close(deserialize_fd(fds, val));
```

yuwata commented on src/core/unit-serialize.c:
```diff
@@ -405,6 +406,98 @@ int unit_deserialize_state_skip(FILE *f) {
+                        int fd;
+
+                        r = extract_first_word(&val, &fdv, NULL, 0);
+                        if (r > 0) {
```
Not necessary to store error code:
```c
if (extract_first_word(...) > 0)
        safe_close(deserialize_fd(...));
```

yuwata commented on src/core/unit-serialize.c:
```diff
@@ -405,6 +406,98 @@ int unit_deserialize_state_skip(FILE *f) {
+                                int fd;
+
+                                r = extract_first_word(&e, &fdstr, ":", 0);
+                                if (r > 0) {
```
Also here.

yuwata commented on src/core/unit-serialize.c:
```diff
@@ -405,6 +406,98 @@ int unit_deserialize_state_skip(FILE *f) {
+                        return 0;
+
+                if (isempty(line))
+                        return 1;
```
Seems to not necessary to return `1`. The caller only checks if it is negative or not. Please use `0`.

yuwata commented on test/units/TEST-07-PID1.alias-corruption.sh:
```diff
@@ -0,0 +1,142 @@
+            systemctl daemon-reload
+
+            # Ensure they are running again (they might have been
+            # abandoned/killed during the transition)
```
When a unit is overridden, the serialization is discarded and the service process is killed, maybe better to check that?

How about to save (not only legit one but also) all sus service PIDs before daemon-reload, and check the consistency between journal entries of 'Skipping stale state...' and killed processes ?

yuwata commented on test/units/TEST-07-PID1.alias-corruption.sh:
```diff
@@ -0,0 +1,142 @@
+
+setup_test_units
+trigger_bug
+verify_result
```
I do not think it is necessary to split the test into three sequential functions.
Please merge them.

Also, maybe better to test with `daemon-reexec`?

yuwata reviewed (commented):
I guess this works fine. Several suggestions.

yuwata reviewed (approved):
LGTM.

keszybz commented on test/units/TEST-07-PID1.alias-corruption.sh:
```diff
@@ -0,0 +1,177 @@
+    echo "Testing with: systemctl $reload_cmd"
+    echo "========================================="
+
+    cat > /run/systemd/system/legit.service <<'EOF'
```
Please drop the space after a redirection operator. Below too.

keszybz commented on src/core/manager-serialize.c:
```diff
@@ -204,6 +204,36 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
+                            name,
+                            u->id);
+
+                return unit_deserialize_state_discard(f, fds);
```
`manager_deserialize_units` calls `manager_deserialize_one_unit` and then `unit_deserialize_state_skip`. This patch introduces `unit_deserialize_state_discard` which also skips data for a unit, except that it doesn't make any attempt to close fds. I think the attempt to add the fd closing here is going to create bugs in the future, because we'll almost certainly forget to update this side list. So for maintainability, we should just handle this in a generic way.

We need to handle left-over fds somehow anyway, and as @yuwata said, we are supposed to be doing that. For example, we might deserialize data from a *newer* systemd, with fields that this copy does not even know about, so we must gracefully support the case where extraneous fds are provided.

That said, I think it'd make sense to set a flag like `boo_stale_aliased_units_have_skipped_in_deserialization = true`, and use that to emit a better message where we warn about unused fds. (Not sure if we do that anywhere.)

EDIT: I guess this wasn't clear what I mean here. Please kill  `unit_deserialize_state_discard` and let `unit_deserialize_state_skip` do the job.

keszybz reviewed (changes requested):
Overall, I think the behaviour after this patch is better than status quo. Corrupting the state of an unrelated unit by merging with state of an unrelated unit is obviously terrible.

OTOH, leaving behind the processes without supervision seems very wrong too. Maybe we could create a temporary unit that'd take over the cgroup so that we at least keep track? And immediately schedule this unit to be killed? I don't know how to do this, and I don't think this should be considered a blocker for merging. Emitting the warning is good enough and better than status quo.

yuwata commented on src/core/main.c:
```diff
@@ -2137,6 +2137,8 @@ static int do_reexecute(
          * but let's hope that doesn't matter.) */
 
         arg_serialization = safe_fclose(arg_serialization);
+        if (fds && !fdset_isempty(fds))
```
`fdset_isempty()` can take `NULL`. Please drop the first condition `fds`.

yuwata commented on src/core/manager-serialize.c:
```diff
@@ -203,6 +203,36 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
+                /*
+                 * The unit from the state file (name) resolved to a different canonical unit (u->id). This
+                 * means the unit in the state file is now an alias that wasn't an alias when the state was
+                 * serialised.
```
Typo: serialized

yuwata commented on src/core/main.c:
```diff
@@ -2137,6 +2137,8 @@ static int do_reexecute(
 
         arg_serialization = safe_fclose(arg_serialization);
+        if (fds && !fdset_isempty(fds))
+                log_warning("%u unused deserialisation file descriptors passed, closing.", fdset_size(fds));
```
deserialization

yuwata commented on src/core/manager-serialize.c:
```diff
@@ -203,6 +203,36 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
+                 * were independent when the state file was written, but are now aliases (either because a
+                 * reload created the symlink, or the symlink existed but this is the first reload). Normal
+                 * aliases that were already aliases during the most recent serialisation are filtered out in
+                 * in manager_serialise_units(), so they never appear in the state file.
```
manager_serialize_units()

yuwata commented on src/core/manager-serialize.c:
```diff
@@ -203,6 +203,36 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
+                 * aliases that were already aliases during the most recent serialisation are filtered out in
+                 * in manager_serialise_units(), so they never appear in the state file.
+                 *
+                 * The serialised data represents the old, independent unit. Deserialising this stale state
```
serialized
Deserializing

yuwata commented on test/units/TEST-07-PID1.alias-corruption.sh:
```diff
@@ -0,0 +1,177 @@
+
+    # Create 20 sus units. They must be Type=simple/running so systemd
+    # CANNOT garbage collect them. If they are dead/stopped, systemd can remove
+    # them from memory before serialisation
```
Also here: serialization

yuwata reviewed (approved):
LGTM.

Please convert all serialise -> serialize, including commit message.

YHNdnzj commented on src/core/main.c:
```diff
@@ -2137,6 +2137,8 @@ static int do_reexecute(
 
         arg_serialization = safe_fclose(arg_serialization);
+        if (fds && !fdset_isempty(fds))
+                log_warning("%u unused deserialisation file descriptors passed, closing.", fdset_size(fds));
```
note that this is very much expected, e.g. we don't consume the serialized varlink socket fds on daemon-reload. Kinda tempted to drop this...

yuwata replied:
Maybe better to downgrade to LOG_DEBUG ?

cdown replied:
That's what I meant above by saying not all file descriptors are even intended to be consumed. I don't have a strong preference here, I don't think it's necessarily very useful.

YHNdnzj replied:
fdset_close debug logs internally already, hence let's not duplicate it.

YHNdnzj reviewed (changes requested):
frankly, I'm a bit skeptical about this approach, considering this might break legitimate use cases. Let's say if a distro packages a.service, and the upstream of "a" got renamed to "b" and the canonical unit of that package is now b.service with an alias a.service, and we'd lose all the states across daemon-reload?

I can see how the simplicity of the approach taken here appeals, but OTOH it feels too naive...

keszybz commented on src/core/manager-serialize.c:
```diff
@@ -207,6 +250,45 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
+                 * The unit from the state file (name) resolved to a different canonical unit (u->id), and
+                 * the canonical unit also has its own state entry.
+                 *
+                 * This means the unit in the state file is now a stale alias that wasn't an alias when the
```
WDYM by "stale alias"? It *wasn't* an alias before, and *is* an alias now. So it's not "stale", IIUC?

cdown replied:
It's a fair question, by this I meant the serialised state entry is stale rather than the alias, in that it was written when it was still an independent unit, but now it resolves an alias to `u->id`. I'll rephrase it overall to:

> This means the state entry for the unit name is stale. That is, when the state was serialized, the name referred to an independent unit, but it now resolves as an alias to the canonical unit.

keszybz reviewed (commented):
LGTM, but I'm confused by the "stale alias" phrasing.

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -197,7 +198,49 @@ int manager_serialize(
+                if (r == 0)
+                        break;
+
+                r = set_put_strdup(&serialized_units, line);
```
Why strdup? Just pass ownership over?

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -197,7 +198,49 @@ int manager_serialize(
 }
 
-static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, FDSet *fds) {
+static int manager_collect_serialized_unit_names(FILE *f, Set **ret_serialized_units) {
```
Calling it ret would be fine. The function name already entails the info

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -207,6 +250,46 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
                 return log_notice_errno(r, "Failed to load unit \"%s\", skipping deserialization: %m", name);
 
+        if (!streq(u->id, name) &&
+            serialized_units &&
```
set_contains handles NULL set gracefully, hence this is unnecessary

cdown committed:
core: Prevent corrupting units from stale alias state on daemon-reload

During daemon-reload (or daemon-reexec), when a unit becomes an alias to
another unit, deserialising the alias's stale serialised state can
corrupt the canonical unit's live runtime state.

Consider this scenario:

1. Before reload:
   - a.service is running
   - b.service was stopped earlier and is dead
   - Both exist as independent units

2. User creates an alias to migrate from b -> a:

   - `ln -s /run/systemd/system/a.service /etc/systemd/system/b.service`

3. daemon-reload triggers serialisation. State file contains both units:
   - a.service -> state=running, cgroup=/system.slice/a.service,
     PID=1234, ...
   - b.service -> state=dead, cgroup=(empty), no PIDs, ...

4. During deserialisation:
   - Processes a.service: loads Unit A, deserialises -> state=RUNNING
   - Processes b.service: manager_load_unit() detects symlink, returns
     Unit A
   - unit_deserialize_state(Unit A, ...) overwrites with b's dead state

5. The result is that:
   - Unit A incorrectly shows state=dead despite PID 1234 still running
   - If a.service has Upholds= dependents, catch-up logic sees
     a.service should be running but is dead
   - systemd starts a.service again -> PID 5678
   - Two instances run: PID 1234 (left-over) and PID 5678 (new)

This bug is deterministic when serialisation orders a.service before
b.service.

The root cause is that manager_deserialize_one_unit() calls
manager_load_unit(name, &u) which resolves aliases via
unit_follow_merge(), returning the canonical Unit object. However, the
code doesn't distinguish between two cases when u->id differs from the
requested name from the state file. In the corruption case, we're
deserialising an alias entry and unit_deserialize_state() blindly
overwrites the canonical unit's fields with stale data from the old,
independent unit. The serialised b.service then overwrites Unit A's
correct live state.

This commit first scans the serialised unit names, then adds a check
after manager_load_unit():

    if (!streq(u->id, name) && set_contains(serialized_units, u->id))
        ...

This detects when the loaded unit's canonical ID (u->id) differs from
the serialised name, indicating the name is now an alias for a different
unit and the canonical unit also has its own serialised state entry.

If the canonical unit does not have its own serialised state entry, we
keep the state entry. That handles cases where the old name is really
just a rename, and thus the old name is the only serialised state for
the unit. In that case there is no bug, because there is no separate
canonical state entry for the stale alias entry to overwrite.

Skipping is safe because:

1. The canonical unit's own state entry will be correctly deserialised
   regardless of order. This fix only prevents other stale alias entries
   from corrupting it.
2. unit_merge() has already transferred the necessary data. When
   b.service became an alias during unit loading, unit_merge() already
   migrated dependencies and references to the canonical unit.
3. After merging, the alias doesn't have its own runtime state. The
   serialised data represents b.service when it was independent, which
   is now obsolete once the canonical unit also has its own serialised
   entry.
4. All fields are stale. unit_deserialize_state() would overwrite state,
   timestamps, cgroup paths, pids, etc. There's no scenario where we
   want this data applied on top of the canonical unit's own serialised
   state.

This fix also correctly handles unit precedence. For example, imagine
this scenario:

1. `b.service` is a valid, running unit defined in `/run`.
2. The sysadmin creates `ln -s .../a.service /etc/.../b.service`.
3. On reload, the new symlink in `/etc` overrides the unit in `/run`.

The new perspective from the manager side is that `b.service` is now an
alias for `a.service`.

In this case, systemd correctly abandons the old b.service unit, because
that's the intended general semantics of unit file precedence. We also
do that in other cases, like when a unit file in /etc/systemd/system/
masks a vendor-supplied unit file in /lib/systemd/system/, or when an
admin uses systemctl mask to explicitly disable a unit.

In all these scenarios, the configuration with the highest precedence
(in /etc/) is treated as the new source of truth. The old unit's
definition is discarded, and its running processes are (correctly)
abandoned. In that respect we are not doing anything new here.

Some may ask why we shouldn't just ignore the symlink if we think this
case will come up. I think there are multiple very strong reasons not to
do so:

1. It violates unit precedence. The unit design is built on a strict
   precedence list. When an admin puts any file in /etc, they are
   intentionally overriding everything else. If manager_load_unit were
   to "ignore" this file based on runtime state, it would break this
   fundamental precedent.
2. It makes daemon-reload stateful. daemon-reload is supposed to be a
   simple, stateless operation, basically to read the files on disk and
   apply the new configuration. But doing this would make daemon-reload
   stateful, because we'd have to read the files on disk, but
   cross-reference the current runtime state, and...  maybe ignore some
   files. This is complex and unpredictable.
3. It also completely ignores the user intent. The admin clearly has
   tried to replace the old service with an alias. Ignoring their
   instruction is the opposite of what they want.

Fixes: https://github.com/systemd/systemd/issues/38817
Fixes: https://github.com/systemd/systemd/issues/37482

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -197,7 +198,49 @@ int manager_serialize(
+                return log_error_errno(errno, "Failed to determine serialization offset: %m");
+
+        for (;;) {
+                char *line = NULL;
```
initialization is unnecessary

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -197,7 +198,49 @@ int manager_serialize(
+
+                r = set_ensure_consume(&serialized_units, &string_hash_ops_free, line);
+                if (r < 0)
+                        return r;
```
log_oom?

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -197,7 +198,49 @@ int manager_serialize(
+                const char *name,
+                FILE *f,
+                FDSet *fds,
+                const Set *serialized_units) {
```
Please insert a newline, as per our coding style

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -216,7 +298,11 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
+                Manager *m,
+                FILE *f,
+                FDSet *fds,
+                const Set *serialized_units) {
```
ditto

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -207,6 +250,45 @@ static int manager_deserialize_one_unit(Manager *m, const char *name, FILE *f, F
                 return log_notice_errno(r, "Failed to load unit \"%s\", skipping deserialization: %m", name);
 
+        if (!streq(u->id, name) &&
+            set_contains(serialized_units, u->id)) {
```
Hmm, this misses the (even more unlikely) case where the new canonical unit id was not present, but _multiple_ individual units got turned into aliases of it? I figure we should insert the canonical unit id to the set unconditionally after deserialization to ensure we skip further attempts to deserialize to the unit object

YHNdnzj commented on src/core/manager-serialize.c:
```diff
@@ -555,5 +642,9 @@ int manager_deserialize(Manager *m, FILE *f, FDSet *fds) {
         }
 
-        return manager_deserialize_units(m, f, fds);
+        r = manager_collect_serialized_unit_names(f, &serialized_units);
```
Hmm, why not fold this into _deserialize_units?

YHNdnzj reviewed (approved):
Mostly LGTM, thanks!

ikruglov commented on src/core/manager-serialize.c:
```diff
@@ -197,7 +198,50 @@ int manager_serialize(
+                return log_error_errno(errno, "Failed to determine serialization offset: %m");
+
+        for (;;) {
+                char *line;
```
should be `_cleanup_free_ char *line = NULL;`

cdown replied:
Good catch! The fix will have to be a little more complicated than just changing it to `_cleanup_free_ char *line = NULL;` , since we also need to TAKE_PTR(line) into set_ensure_consume(). I'll update that now.

ikruglov reviewed (changes requested):
ah, there is a memleak. @cdown please have a look

cdown commented on test/units/TEST-07-PID1.alias-corruption.sh:
```diff
@@ -0,0 +1,178 @@
+        # Check consistency between journal warnings and abandoned processes
+        echo "Checking journal for stale state warnings..."
+        journal_warnings=$(journalctl --since "$reload_start" --no-pager | grep "Skipping stale state" || true)
+        warning_count=$(echo "$journal_warnings" | grep -c "Skipping stale state" || true)
```
Because we don't want false failures and this is probabilistic, this is actually what we want. So I don't think we should change this.

poettering commented on src/core/manager-serialize.c:
```diff
@@ -197,7 +198,50 @@ int manager_serialize(
+        }
+
+        if (fseeko(f, offset, SEEK_SET) < 0)
+                return log_error_errno(errno, "Failed to reset serialization offset: %m");
```
urks, i cannot say i find an approach that parses forwards and returns to the beginning particularly sexy...
